### PR TITLE
Confirm deposit service

### LINKED ISSUE #300: Implement ConfirmDeposit service
Confirm deposit
- Generate proof for deposit message declaration
- Generate raw transaction for confirm deposit
- Call transaction executor

### PR DESCRIPTION
**Pull request by 0xsarvesh**
fixes #300

### CODE REVIEW

File: src/m1_facilitator/services/ConfirmDepositService.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+  private async confirmDepositIntentTransaction(
+    message: Message,
+    depositIntent: DepositIntent,
+    gateway: Gateway,
+  ): Promise<TransactionObject<string>> {
+    const cogatewayAddress = gateway.remoteGA;
+    const erc20Cogateway = Mosaic.interacts.getERC20Cogateway(this.auxiliaryWeb3, cogatewayAddress);
+
+    const blockNumber = gateway.remoteGatewayLastProvenBlockNumber;
+    const proof = await new ProofGenerator(this.originWeb3).getOutboxProof(
```
Comment (0xsarvesh) [inline src/m1_facilitator/services/ConfirmDepositService.ts:143]: pass the storage index.

File: src/m1_facilitator/repositories/MessageRepository.ts [hunk 2]
```diff
@@ -216,6 +218,43 @@ export default class MessageRepository extends Subject<Message> {
     return this.convertToMessage(messageModel);
   }
 
+  /**
+   * This method returns messages based on below criteria.
+   *   - Filter based  on gateway address.
```
Comment (deepesh-kn) [inline src/m1_facilitator/repositories/MessageRepository.ts]:
```suggestion
   *   - Filter based on gateway address.
```

File: src/m1_facilitator/services/ConfirmDepositService.ts (hunk 1 above)
Comment (deepesh-kn) [reply src/m1_facilitator/services/ConfirmDepositService.ts:143]: The proof generator may not work as per the expectations.
The path formation is different for `Mosaic-1` contracts

File: src/m1_facilitator/services/ConfirmDepositService.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+   *     to remote gateway last proven block height.
+   *  3. Target status of message should be undeclared.
+   *  4. Message should be initiated from given gateway.
+   *
+   * If there are pending messages based on above criteria, confirm deposit
+   * transaction is sent.
+   *
+   * @param gateways List of gateway object.
+   */
+  public async update(gateways: Gateway[]) {
```
Comment (deepesh-kn) [inline src/m1_facilitator/services/ConfirmDepositService.ts]:
```suggestion
  public async update(gateways: Gateway[]): Promise<void> {
```

File: test/m1_facilitator/repositories/MessageRepository/getPendingMessagesByGateway.test.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  from '../../../../src/m1_facilitator/repositories/Repositories';
+import MessageRepository
+  from '../../../../src/m1_facilitator/repositories/MessageRepository';
+import Message, {
+  MessageStatus,
+  MessageType,
+} from '../../../../src/m1_facilitator/models/Message';
+import assert from '../../../test_utils/assert';
+
+describe('MessageRepository:getPendingMessagesByGateway', () => {
```
Comment (deepesh-kn) [inline test/m1_facilitator/repositories/MessageRepository/getPendingMessagesByGateway.test.ts]:
```suggestion
describe('MessageRepository:getPendingMessagesByGateway', (): void  => {
```

File: test/m1_facilitator/services/ConfirmDepositService/update.test.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  MessageStatus,
+  MessageType,
+} from '../../../../src/m1_facilitator/models/Message';
+import DepositIntent
+  from '../../../../src/m1_facilitator/models/DepositIntent';
+import SpyAssert from '../../../test_utils/SpyAssert';
+import TransactionExecutor
+  from '../../../../src/m1_facilitator/lib/TransactionExecutor';
+
+describe('ConfirmDepositService:update ', () => {
```
Comment (deepesh-kn) [inline test/m1_facilitator/services/ConfirmDepositService/update.test.ts]:
```suggestion
describe('ConfirmDepositService:update ', (): void => {
```

File: test/m1_facilitator/services/ConfirmDepositService/update.test.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+  let depositIntent: DepositIntent;
+  const confirmDepositRawTx = 'Some raw tx';
+  const serializedProof = 'Storage proof';
+  let transactionExecutor: any;
+  let fakeERC20Cogateway: any;
+  let getERC20CogatewaySpy: any;
+  let auxiliaryWeb3: any;
+  let confirmDepositSpy: any;
+
+  beforeEach(async () => {
```
Comment (deepesh-kn) [inline test/m1_facilitator/services/ConfirmDepositService/update.test.ts]:
```suggestion
  beforeEach(async (): Promise<void> => {
```

File: test/m1_facilitator/services/ConfirmDepositService/update.test.ts [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  let gateway: Gateway;
+  let message: Message;
+  let depositIntent: DepositIntent;
+  const confirmDepositRawTx = 'Some raw tx';
+  const serializedProof = 'Storage proof';
+  let transactionExecutor: any;
+  let fakeERC20Cogateway: any;
+  let getERC20CogatewaySpy: any;
+  let auxiliaryWeb3: any;
+  let confirmDepositSpy: any;
```
Comment (deepesh-kn) [inline test/m1_facilitator/services/ConfirmDepositService/update.test.ts:53]: Disable the linting errors.
```
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let transactionExecutor: any;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let fakeERC20Cogateway: any;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let getERC20CogatewaySpy: any;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let auxiliaryWeb3: any;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let confirmDepositSpy: any;

```

File: test/m1_facilitator/services/ConfirmDepositService/update.test.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+    await repositories.depositIntentRepository.save(depositIntent);
+    await repositories.messageRepository.save(message);
+
+    const originWeb3 = sinon.createStubInstance(Web3);
+    auxiliaryWeb3 = sinon.createStubInstance(Web3);
+
+    transactionExecutor = sinon.createStubInstance(TransactionExecutor);
+
+    fakeERC20Cogateway = {
```
Comment (deepesh-kn) [inline test/m1_facilitator/services/ConfirmDepositService/update.test.ts:96]: Please fix the linting issue:
`Missing return type on function.eslint(@typescript-eslint/explicit-function-return-type)`

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
Looks good to me,  just  a few linting issues

File: src/m1_facilitator/repositories/MessageRepository.ts (hunk 2 above)
Comment (0xsarvesh) [reply src/m1_facilitator/repositories/MessageRepository.ts]: Done

File: src/m1_facilitator/services/ConfirmDepositService.ts (hunk 3 above)
Comment (0xsarvesh) [reply src/m1_facilitator/services/ConfirmDepositService.ts]: Done

File: test/m1_facilitator/repositories/MessageRepository/getPendingMessagesByGateway.test.ts (hunk 4 above)
Comment (0xsarvesh) [reply test/m1_facilitator/repositories/MessageRepository/getPendingMessagesByGateway.test.ts]: Done

File: test/m1_facilitator/services/ConfirmDepositService/update.test.ts (hunk 5 above)
Comment (0xsarvesh) [reply test/m1_facilitator/services/ConfirmDepositService/update.test.ts]: Done

File: test/m1_facilitator/services/ConfirmDepositService/update.test.ts (hunk 6 above)
Comment (0xsarvesh) [reply test/m1_facilitator/services/ConfirmDepositService/update.test.ts]: Done

File: test/m1_facilitator/services/ConfirmDepositService/update.test.ts (hunk 7 above)
Comment (0xsarvesh) [reply test/m1_facilitator/services/ConfirmDepositService/update.test.ts:53]: Done

File: test/m1_facilitator/services/ConfirmDepositService/update.test.ts (hunk 8 above)
Comment (0xsarvesh) [reply test/m1_facilitator/services/ConfirmDepositService/update.test.ts:96]: Done

File: src/m1_facilitator/services/ConfirmDepositService.ts (hunk 1 above)
Comment (deepesh-kn) [reply src/m1_facilitator/services/ConfirmDepositService.ts:143]: Merging this, I will create a ticket to fix it in `@openst/mosaic-proof`

**Review by deepesh-kn — APPROVED** [review]

**Pull request merged**